### DISCUSSION TITLE
Comments on State Logic

### DISCUSSION THREAD

**Discussion by musslick** [AutoResearch/autora#543] [Ideas]
This is a thread to collect comments/suggestions on using the latest state logic in https://github.com/AutoResearch/autora-core/tree/feat/mini-hackathon/docs/cycle

**Comment by musslick** [comment]
**FEATURE REQUEST**: ``from autora.state.bundled import StandardState`` -- What does ``bundled`` mean? Would be more intuitive to import ``StatdardState`` directly from ``autora.state``.

**Reply by musslick** [reply]
This should be fixed with the new refactoring.
We renamed 
`autora.state.bundled` -> `autora.state.standard` and
`autora.state.delta` -> autora.state.__init__

**Reply by hollandjg** [reply]
I think it would be good to keep the `delta` stuff as its own sub package for now. What we could do is say:
- If you want fundamental things with the State/Delta mechanism, those can be imported from `autora.state.delta`, 
- if you want the "standard" state things like the `StandardState` and wrappers `state_fn_from_estimator`..., then you can import those from `autora.state.standard`

(moving the things from `delta` to \_\_init\_\_ can't work, unfortunately, as that would break the namespace-package structure and make it more difficult to add a new way of doing things to the `autora.state` sub package in future)

**Reply by musslick** [reply]
I think the term "delta" is foreign to most users and importing state from it does not feel intuitive. I would strongly advocate for something simpler "state". We should discuss this.

**Reply by hollandjg** [reply]
@musslick , @younesStrittmatter and I talked about:
- [x] merge all the state things we still use into a single file state.py https://github.com/AutoResearch/autora-core/pull/33
- [x] delete the old state objects (Snapshot, History) – those are only used by the old workflow logic which needs completely rewriting. https://github.com/AutoResearch/autora-core/pull/33
- [x] Update current version of autora-workflow to specify aurora-core < 3.0.0 https://github.com/AutoResearch/autora-workflow/pull/38

**Comment by musslick** [comment]
**FEATURE REQUEST**: append logic for VariablesCollection

**Reply by hollandjg** [reply]
I'm not sure how this would work. The first example we made where we needed this was actually a deletion – how would you want the interface to look to deal with deletions, replacements and additions?

**Reply by musslick** [reply]
Let's make this lower priority for now. 
I'm worried that the `+` operator is not intuitive to deletion. It should be an addition by default.

**Comment by musslick** [comment]
**FEATURE REQUEST:** Be able to declare input variables in the wrapper function

**Reply by hollandjg** [reply]
How do you mean?

**Reply by musslick** [reply]
E.g., specify not only the outputs but the inputs when calling `on_state`. Tagging @younesStrittmatter here who came up with the suggestion.

**Comment by musslick** [comment]
**FEATURE REQUEST:** when creating new state, make it more convenient to add fields to the state. It's inconvenient now because you have to create a new state class.

**Reply by hollandjg** [reply]
This isn't possible using dataclasses as far as I know – you always have to create a new state class. 
What someone could do is implement the same logic but using a UserDict – they would have to work out how to translate what we've done for data classes so that it works on a dictionary, but that would be totally doable.

**Reply by hollandjg** [reply]
This needs to be a separate issue

**Reply by musslick** [reply]
Sounds good, also tagging @younesStrittmatter here who raised this issue.

**Comment by musslick** [comment]
**BUG:**  model list problem @younesStrittmatter

**Reply by chadcwilliams** [reply]
Is this that the entire State model list represents the last model and not the history of models? 

If so, +1.

**Reply by hollandjg** [reply]
Please can someone add a bit more info here? We can turn this into an issue.

**Reply by younesStrittmatter** [reply]
The issue is: When creating a custom State with fields as lists, then the Delta Object needs to return a list as well:
class ExtendedState(StandardState):
    models_bms: List[BaseEstimator] = field(
        default_factory=list,
        metadata={"delta": "extend"},
    )
    
   @on_state()
def bms_theorist(experiment_data: pd.DataFrame, variables: VariableCollection, **kwargs):
    ivs = [v.name for v in variables.independent_variables]
    dvs = [v.name for v in variables.dependent_variables]
    X, y = experiment_data[ivs], experiment_data[dvs]
    new_model = BMSRegressor(epochs=EPOCHS).set_params(**kwargs).fit(X, y)
    return Delta(models_bms=[new_model])

**Reply by younesStrittmatter** [reply]
The confusion came from the option to return a model (without the s) with the StandardState that is not a list because it has an alias to do so. So long story short: This might be okay as is and doesn't need to be an issue.

**Reply by musslick** [reply]
enforce that return Delta has to be a list

**Reply by musslick** [reply]
make sure to mention this in documentation

**Comment by chadcwilliams** [comment]
**FEATURE REQUEST:**

`from autora.state.bundled import StandardState`
`from autora.state.delta import on_state`
`from autora.state.wrapper import state_fn_from_estimator`

Definitely not a problem, but I find it odd that these all come from different places, especially the latter two as they do the same thing (conceptually).

It'd be nice to be able to:
`from autora.state.bundled import StandardState, on_state, state_fn_from_estimator`

**Reply by hollandjg** [reply]
I've made an update to [feat: new general wrapper for State mechanism](https://github.com/AutoResearch/autora-core/pull/39) so that you can do 
```python
from autora.state.standard import StandardState, on_state, state_fn_from_estimator
```

**Reply by musslick** [reply]
Strongly agree with chad! Moving these to standard may help here and should resolve the issue about naming above.

**Reply by chadcwilliams** [reply]
Thanks for doing that John!

**Reply by hollandjg** [reply]
Another update: as discussed, I've simplified it again and now it looks like:
```python
from autora.state import StandardState, on_state, state_fn_from_estimator
```
– everything is in one file now.

**Comment by hollandjg** [comment]
Hi all, I've updated https://github.com/AutoResearch/autora-core/pull/33 with all the key changes from the mini hackathon. 

Please note: I didn't include the align_dataframe_to_ivs function from https://github.com/AutoResearch/autora-core/blob/mini-hackaton-before-refactor-state/src/autora/utils/conversion.py because I don't think this should be part of the on_state function. There's an explanation here: https://github.com/AutoResearch/autora-core/pull/33#issuecomment-1688098399

I would like to suggest that we include any additional features or requests as separate PRs. (However, if anything is broken with the existing PR, we'll update that.)

**Comment by musslick** [comment]
**FEATURE REQUEST:** some logic to combine outputs from multiple components into one output. E.g., current problem is that when two experimentalists are chained; the second one overwrites the first. We want some functionality that combines the output of both experimentalists @chadcwilliams

**Reply by chadcwilliams** [reply]
Now, this is going to be asking too much and definitely not necessary, but it would be cool if we can add this functionality without adding new wrapping functions (e.g., a wrapping function to combine two experimentalists). It would be most intuitive, in my opinion, to be able to simply write:

```
s = experimentalist1(s, num_samples=10)
s = experimentalist2(s, num_samples=10)
s = experiment_runner(s)
s = theorist(s)
```

One potential way of doing this is to keep a history of what has been run (which, I think might be a good idea anyhow). Something like a list that could look like:

```
[EXPERIMENTALIST,
EXPERIMENTALIST,
EXPERIMENT_RUNNER,
THEORIST,
EXPERIMENTALIST,
EXPERIMENTALIST,
EXPERIMENT_RUNNER,
THEORIST]
```
If we have this history being tracked, when an experimentalist acts on the state, it can look back through the history and see if a different experimentalist was run since the last runner or not. If yes another experimentalist has been run since the last runner, then we concatenate the current experimentalist conditions with the state conditions. If no another experimentalist has **not** been run since the last runner, then we replace conditions as normal. We could do something similar with theorists where we can see how many theorists were run between experimentalists.

Potential example method to searching the history:
```
def walk_history(self):
    for component in reversed(self.history):
        if (component =='EXPERIMENTALIST') | (component =='EXPERIMENT_RUNNER'):
            return component
    
if walk_history(self) == 'EXPERIMENTALIST':
    #concatenate conditions
else: 
    #replace conditions
```

But, I haven't thought this through too thoroughly, so maybe there's a lot of room for bugs using this type of approach.

**Reply by younesStrittmatter** [reply]
I think this is not a good solution. For example: What if our first experimentalist is a pooler and the second one is a sampler (a quite likely scenario).

**Reply by chadcwilliams** [reply]
In this scenario, does the first pooler experimentalist refer to creating a condition pool/ condition seed? We could always avoid looking at the first entry if a seed is provided. 

Otherwise, I agree adding a pooler result to a sampler result wouldn't be the functionality we expect to endorse, but that could be up to the user.

**Reply by younesStrittmatter** [reply]
That was the idea behind this method of chaining anyway. We have a Pipeline of experimentalists that the conditions are piped through. Maybe in the old way of doing this, it was more obvious, since all of the experimentalist samplers used conditions as input. Only the poolers "create" new conditions. In your example, the 2nd experimentalist would use the output of the first experimentalist as pool and then append it to the conditions that are already there. I don't think that is what we want. I think the most clean way of doing this is to create a new field in the state, name it "pool" and use the pool for both experimentalist as input. The experimentalists then output the conditions into their own fields "condition_exp1", "condition_exp2" Then create a new experimentalist that combines both of the outputs of the experimentalists together in one "conditions" field.

**Reply by younesStrittmatter** [reply]
That is, if we want to use state logic. You can also use your own function wrapped around the two experimentalist: 
@on_state()
def example(conditions):
      cond_1 = experimentalist_1(conditions)
      cond_2 = experimentalist_2(conditions)
      return Delta("conditions":pd.concat([cond_1, cond_2), axis=0))

**Reply by younesStrittmatter** [reply]
The poolers and samplers both write into a single "conditions" field of the state for now. The difference of the both is, that the poolers use the "variables" field as input and the samplers use the "conditions" field as an input

**Reply by younesStrittmatter** [reply]
BTW: That is something, we should make very clear. If you only use samplers in your cycle, that means, you always use the conditions from your last cycle as a pool for the samplers. So your first experimentalist should almost always be a pooler.

**Reply by younesStrittmatter** [reply]
*first experimentalist in each cycle

**Reply by chadcwilliams** [reply]
Oh, weird, I didn't realize it used the previous cycles conditions. But regardless, I'm happy with any implementation.

**Reply by hollandjg** [reply]
This probably needs more discussion, and potentially an exploration of the best way to chain experimentalists. 

The reason that the we replace the conditions rather than appending to them is that we said "it's weird to append conditions, each time round you actually create some new ones and only use the newest ones, and so the conditions need to be replaced." This can be a thing for the next major version – updating the way we handle this in the next major version.

**Reply by musslick** [reply]
should cover this in the tutorial

**Reply by musslick** [reply]
potentially add some functionality to ease tutorial

**Comment by musslick** [comment]
**DECISION**: Have experimentalists & runners accept pandas df, (+numpy)  and return pandas df
